src/net/download.rs

```diff
@@ -0,0 +1,1904 @@
+
+        Ok(blocks_to_try)
+    }
+
+    /// Make requests for missing anchored blocks
+    fn make_block_requests(&mut self, burndb: &mut BurnDB, chainstate: &mut StacksChainState, start_sortition_height: u64) -> Result<HashMap<u64, VecDeque<BlockRequestKey>>, net_error> {
+        self.make_requests(burndb, chainstate, start_sortition_height, false)
+    }
+
+    /// Make requests for missing confirmed microblocks 
+    fn make_confirmed_microblock_requests(&mut self, burndb: &mut BurnDB, chainstate: &mut StacksChainState, start_sortition_height: u64) -> Result<HashMap<u64, VecDeque<BlockRequestKey>>, net_error> {
+        self.make_requests(burndb, chainstate, start_sortition_height, true)
+    }
+
+    /// Prioritize block requests -- ask for the rarest blocks first
+    fn prioritize_requests(requests: &HashMap<u64, VecDeque<BlockRequestKey>>) -> Vec<u64> {
+        let mut ordered = vec![];
+        for (block_height, requests) in requests.iter() {
+            ordered.push((*block_height, requests.len()));
+        }
+        ordered.sort_by(|(_, ref l1), (_, ref l2)| l1.cmp(l2));
+        ordered.iter().map(|(ref h, _)| *h).collect()
+    }
+
+    /// Go start resolving block URLs to their IP addresses
+    pub fn block_dns_lookups_begin(&mut self, burndb: &mut BurnDB, chainstate: &mut StacksChainState, dns_client: &mut DNSClient) -> Result<(), net_error> {
+        test_debug!("{:?}: block_dns_lookups_begin", &self.local_peer);
+        let (need_blocks, block_sortition_height, microblock_sortition_height) = match self.block_downloader {
+            Some(ref mut downloader) => (downloader.blocks_to_try.len() == 0, downloader.block_sortition_height, downloader.microblock_sortition_height),
+            None => {
+                test_debug!("{:?}: downloader not connected", &self.local_peer);
+                return Err(net_error::NotConnected);
+            }
+        };
+
+        if need_blocks {
+            PeerNetwork::with_downloader_state(self, |ref mut network, ref mut downloader| {
+                test_debug!("{:?}: needs blocks", &network.local_peer);
+
+                let mut next_block_sortition_height = block_sortition_height;
+                let mut next_microblock_sortition_height = microblock_sortition_height;
+
+                test_debug!("{:?}: Look for blocks at sortition {}, microblocks at sortition {}", &network.local_peer, next_block_sortition_height, next_microblock_sortition_height);
+
+                // fetch as many blocks and microblocks as we can -- either
+                // downloader.max_inflight_requests, or however many blocks remain between the
+                // downloader's sortition height and the chain tip's sortition height (whichever is
+                // smaller).
+                while next_block_sortition_height <= network.chain_view.burn_block_height - burndb.first_block_height && next_microblock_sortition_height <= network.chain_view.burn_block_height - burndb.first_block_height {
```

Comment:
Is it possible for next_block_sortition_height and next_microblock_sortition_height to be different ? If so, shouldn't this condition be an 'or' rather than an 'and' ?

Reply:
Yes; in fact, they'll be different more often than not because the downloader can only fetch a microblock stream once it has both the anchored block that produced it and the anchored block that confirms it.  So, the effect is that the block sortition height is usually ahead of the microblock sortition height.

However, it doesn't make a difference in practice, since the calls to `make_block_requests()` and `make_confirmed_microblock_requests()` below will simply return zero entries if the sortition heights exceed the chain height.  But, will change it nevertheless.

Reply:
Fixed in 2fccb98